File: Image 라벨링/c++/labeling/labeling/labeling.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

int main()
{
	Mat img_color = imread("label.jpg", 1);
	Mat img_gray;
	cvtColor(img_color, img_gray, COLOR_BGR2GRAY);
	imshow("result", img_gray);
	waitKey(0);

	// 엣지를 검출 검출된 엣지부분만 흰색
	Mat img_edge;
	Canny(img_gray,img_edge,50, 150);
	imshow("result", img_edge);
	waitKey(0);

	// 관심물체가 흰색이 되어야 하므로 반전
	bitwise_not(img_edge,img_edge);
	imshow("result", img_edge);
	waitKey(0);

	// 컨투어를 찾아서 외곽선을 보강
	vector<vector<Point>> contours;
	findContours(img_edge.clone(), contours, RETR_LIST, CHAIN_APPROX_SIMPLE);
	drawContours(img_edge, contours, -1, Scalar(0, 0, 0), 1);
	imshow("result", img_edge);
	waitKey(0);

	// 흰색 영역에 대해 라벨링
	Mat labels, stats, centroids;
	int nlabels = connectedComponentsWithStats(img_edge, labels, stats, centroids);

	for (int i = 0; i < nlabels; i++)
	{
		// 배경을 제외
		if (i < 2) continue;
		int area = stats.at<int>(i, CC_STAT_AREA);
		int center_x = centroids.at<double>(i, 0);
		int center_y = centroids.at<double>(i, 1);
		int left = stats.at<int>(i, CC_STAT_LEFT);
		int top = stats.at<int>(i, CC_STAT_TOP);
		int width = stats.at<int>(i, CC_STAT_WIDTH);
		int height = stats.at<int>(i, CC_STAT_HEIGHT);

		if (area > 50)
		{
			// 영역 외곽에 사각형을 그립니다.
			rectangle(img_color, Point(left, top), Point(left + width, top + height), Scalar(0, 0, 255), 1);
			// 영역의 중심에 원을 그립니다.
			circle(img_color, Point(center_x, center_y), 5, (255, 0, 0), 1);
			// 라벨 번호 표시
			putText(img_color, std::to_string(i), Point(left + 20, top + 20), FONT_HERSHEY_SIMPLEX, 1, Scalar(255, 0, 0), 2);
		}
	}

	// 결과창
	imshow("result", img_color);
	waitKey(0);
}

File: Image 라벨링/python/Labeling.py
import cv2 as cv


img_color = cv.imread("keyboard.jpeg", cv.IMREAD_COLOR)
img_gray = cv.cvtColor(img_color, cv.COLOR_BGR2GRAY) 
cv.imshow("result", img_gray )
cv.waitKey(0)

#엣지를 검출한다. 검출 된 엣지만 흰색이 됨
img_edge = cv.Canny(img_gray, 50, 150)
cv.imshow("result", img_edge )
cv.waitKey(0)

#관심 물체가 흰색이 되어야 하므로 반전시킨다.
img_edge = cv.bitwise_not(img_edge)
cv.imshow("result", img_edge )
cv.waitKey(0) 

#컨투어를 찾아와서 외곽선을 보강한다.
contours = cv.findContours(img_edge.copy(), cv.RETR_LIST, 
            cv.CHAIN_APPROX_SIMPLE)
cv.drawContours(img_edge, contours[0], -1, (0, 0, 0), 1)
cv.imshow("result", img_edge )
cv.waitKey(0) 

#흰색 영역에 대해 라벨링함.
nlabels, labels, stats, centroids = cv.connectedComponentsWithStats(img_edge)


for i in range(nlabels):

    #배경 제외
    if i < 2:
        continue

    #흰색 영역의 크기, 중심 좌표, 외각 사각형에 대한 정보를 가져온다.
    area = stats[i, cv.CC_STAT_AREA]
    center_x = int(centroids[i, 0])
    center_y = int(centroids[i, 1]) 
    left = stats[i, cv.CC_STAT_LEFT]
    top = stats[i, cv.CC_STAT_TOP]
    width = stats[i, cv.CC_STAT_WIDTH]
    height = stats[i, cv.CC_STAT_HEIGHT]


    #영역의 크기가 50 이상인 경우, 해당 영역에 다음과 같은 정보를 표시.
    if area > 50: 

        #영역 외곽에 사각형을 그림.
        cv.rectangle(img_color, (left, top), (left + width, top + height), (0, 0, 255), 1)
                                                                             ##B,G,R

        #영역 중심 좌표에 원을 그림.
        cv.circle(img_color, (center_x, center_y), 3, (0, 0, 255), 1)

        #라벨 번호를 표시.
        cv.putText(img_color, str(i), (left + 20, top + 20), cv.FONT_HERSHEY_PLAIN, 1, (255,210,0), 1);

        ##############사진, 번호, 위치, 폰트이름, 폰트크기, 폰트색상, 폰트두께

#결과를 화면에 보여줌
cv.imshow("result", img_color)
cv.waitKey(0)